File: dotnet/src/Microsoft.Agents.Protocol.Hosting/Core/IStorage.cs
using System.Threading;
using System.Threading.Tasks;

namespace Microsoft.Agents.Protocol.Hosting.Core;

/// <summary>
/// Interface for thread-scoped state storage.
/// </summary>
public interface IStorage
{
    /// <summary>
    /// Gets a value from storage for a specific thread and key.
    /// </summary>
    /// <typeparam name="T">The type of value to retrieve.</typeparam>
    /// <param name="threadId">The thread ID.</param>
    /// <param name="key">The key.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The value, or default if not found.</returns>
    Task<T?> GetAsync<T>(string threadId, string key, CancellationToken cancellationToken = default);

    /// <summary>
    /// Sets a value in storage for a specific thread and key.
    /// </summary>
    /// <typeparam name="T">The type of value to store.</typeparam>
    /// <param name="threadId">The thread ID.</param>
    /// <param name="key">The key.</param>
    /// <param name="value">The value to store.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    Task SetAsync<T>(string threadId, string key, T value, CancellationToken cancellationToken = default);

    /// <summary>
    /// Deletes a value from storage.
    /// </summary>
    /// <param name="threadId">The thread ID.</param>
    /// <param name="key">The key.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    Task DeleteAsync(string threadId, string key, CancellationToken cancellationToken = default);

    /// <summary>
    /// Gets all keys for a specific thread.
    /// </summary>
    /// <param name="threadId">The thread ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Array of keys.</returns>
    Task<string[]> GetKeysAsync(string threadId, CancellationToken cancellationToken = default);

    /// <summary>
    /// Checks if the storage is healthy and accessible.
    /// </summary>
    /// <returns>True if healthy, false otherwise.</returns>
    Task<bool> CheckHealthAsync();
}

File: typescript/packages/agents-protocol-hosting/src/storage/IStorage.ts
/**
 * Interface for persistent state storage.
 *
 * Implementations must be thread-safe and support the configured
 * consistency model.
 */
export interface IStorage {
  /**
   * Gets a value from storage.
   *
   * @param threadId - The thread ID
   * @param key - The state key
   * @returns The value or null if not found
   */
  getAsync<T>(threadId: string, key: string): Promise<T | null>;

  /**
   * Sets a value in storage.
   *
   * @param threadId - The thread ID
   * @param key - The state key
   * @param value - The value (must be JSON serializable)
   */
  setAsync<T>(threadId: string, key: string, value: T): Promise<void>;

  /**
   * Deletes a value from storage.
   *
   * @param threadId - The thread ID
   * @param key - The state key
   */
  deleteAsync(threadId: string, key: string): Promise<void>;

  /**
   * Gets all keys for a thread.
   *
   * @param threadId - The thread ID
   * @returns Array of keys
   */
  getKeysAsync(threadId: string): Promise<string[]>;

  /**
   * Checks if the storage is healthy.
   *
   * @returns true if healthy, false otherwise
   */
  checkHealth(): Promise<boolean>;
}

/**
 * Configuration for PostgreSQL storage.
 */
export interface PostgresStorageConfig {
  /**
   * Database connection string.
   */
  connectionString: string;

  /**
   * Connection pool configuration.
   */
  pool?: {
    min: number;
    max: number;
    idleTimeoutMillis?: number;
  };

  /**
   * Consistency model for distributed systems.
   *
   * - 'strong': Reads always return latest write
   * - 'eventual': Reads may return stale data
   * - 'causal': Reads respect causal ordering
   *
   * @default 'strong'
   */
  consistency?: 'strong' | 'eventual' | 'causal';

  /**
   * Conflict resolution strategy.
   *
   * - 'last-write-wins': Latest write wins on conflict
   * - 'first-write-wins': First write wins on conflict
   *
   * @default 'last-write-wins'
   */
  conflictResolution?: 'last-write-wins' | 'first-write-wins';
}
